core/src/validator.rs

```diff
@@ -1,256 +1,212 @@
-            .field("validator_type", &self.validator_type)
-            .field("module", &"<Module is truncated>")
+        f.debug_struct("Chain")
+            .field("loaded_validator", &self.loaded_validator)
+            .field("engine", &"<Engine is truncated>")
             .finish()
     }
 }
 
-impl Chain {
-    /// Construct new [`Chain`].
-    pub fn new() -> Self {
-        Self::default()
+impl Clone for Validator {
+    fn clone(&self) -> Self {
+        Self {
+            loaded_validator: RwLock::new(
+                self.loaded_validator
+                    .read()
+                    .expect("Validator is poisoned")
+                    .clone(),
+            ),
+            engine: self.engine.clone(),
+        }
     }
+}
 
-    /// Add new [`Validator`] to the [`Chain`].
-    ///
-    /// Returns `true` if the validator was added
-    /// and `false` if a validator with the same id already exists.
+impl Validator {
+    /// Upgrade runtime validator with the new one.
     ///
     /// # Errors
     ///
-    /// Fails if WASM module loading fails.
-    pub fn add_validator(&self, validator: Validator) -> Result<MustUse<bool>> {
-        use dashmap::mapref::entry::Entry::*;
-
-        let id = validator.id.clone();
-        let Vacant(vacant) = self.all_validators.entry(id.clone()) else {
-            return Ok(MustUse(false));
-        };
-
-        match self
-            .concrete_type_validators
-            .entry(*validator.validator_type())
-        {
-            Occupied(mut occupied) => {
-                occupied.get_mut().push(id);
-            }
-            Vacant(concrete_type_vacant) => {
-                concrete_type_vacant.insert(vec![id]);
-            }
-        }
-
-        let loaded_validator = LoadedValidator {
-            id: validator.id,
-            validator_type: validator.validator_type,
-            module: wasm::load_module(&self.engine, validator.wasm)?,
-        };
-
-        vacant.insert(loaded_validator);
-        Ok(MustUse(true))
-    }
-
-    /// Remove [`Validator`] from the [`Chain`].
-    ///
-    /// Return `true` if the validator was removed
-    /// and `false` if no validator with the given id was found.
-    #[allow(clippy::expect_used)]
-    pub fn remove_validator(&self, id: &Id) -> bool {
-        self.all_validators.get(id).map_or(false, |entry| {
-            let type_ = &entry.validator_type;
-
-            self.all_validators
-                .remove(id)
-                .and_then(|_| self.concrete_type_validators.get_mut(type_))
-                .expect(
-                    "Validator chain internal collections inconsistency error \
-                         when removing a validator. This is a bug",
-                )
-                .retain(|validator_id| validator_id != id);
-            true
-        })
+    /// Fails if there there is an error while loading wasm blob.
+    pub fn upgrade(&self, raw_validator: data_model_validator::Validator) -> Result<()> {
```

Comment:
help me understand. `Upgrade` does not add new validator into the previous monolithic validator, it actually replaces the previous validator. This means that user must make a complete new monolithic validator?

Reply:
Yes, that's it.

In my mind it looks like so:

1. User hase his validator on GitHub
2. When they need to change it, they update the code on GitHub
3. Other developers review it and approve
4. Validator is being compiled to `wasm` and submited as an upgraded version of a previous one

Reply:
we can close this discussion, but your comment is so useful that we can leave it hanging there for other reviewers

Reply:
Created #3375